leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```diff
@@ -178,14 +178,21 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
         try {
             return method.invoke(delegate, args);
         } catch (InvocationTargetException e) {
-            if (e.getCause() instanceof ServiceNotAvailableException
-                    || e.getCause() instanceof NotCurrentLeaderException) {
+            if (e.getTargetException() instanceof ServiceNotAvailableException
+                    || e.getTargetException() instanceof NotCurrentLeaderException) {
                 markAsNotLeading(leadershipToken, e.getCause());
             }
-            throw e.getCause();
+            if (e.getTargetException() instanceof InterruptedException && !(checkIfLeading(leadershipToken))) {
+                throw notCurrentLeaderException("received an interrupt due to leader election.", e.getTargetException());
+            }
+            throw e.getTargetException();
         }
     }
 
+    private boolean checkIfLeading(LeadershipToken leadershipToken) {
+        return leaderElectionService.isStillLeading(leadershipToken) == StillLeadingStatus.LEADING;
```

Comment:
This will mean the proxy will not be aware until a second call received a 503 and makes the leadershipToken as null. The leadership token is changed when someone `gainsLeadership` or a request receives ServiceUnavailable/NCLE. So, if we check the `leadershipTokenRef ` all lock requests will 500 until another request hits and 503. :(